website/docs/internals/provider-network-mirror-protocol.html.md

```diff
@@ -0,0 +1,278 @@
+sensitive then it must use cryptographically-secure, user-specific, and
+time-limited URLs in the metadata response. Strategies for doing so are out
+of scope of this protocol documentation.
+
+## List Available Versions
+
+This operation determines which versions are currently available for a
+particular provider.
+
+| Method | Path                                    | Produces           |
+|--------|-----------------------------------------|--------------------|
+| `GET`  | `:hostname/:namespace/:type/index.json` | `application/json` |
+
+### Parameters
+
+* `hostname` (required): the hostname portion of the address of the requested
+  provider.
+* `namespace` (required): the namespace portion of the address of the requested
+  provider.
+* `type` (required): the type portion of the address of the requested provider.
+
+### Sample Request
+
+```
+curl 'https://terraform.example.com/providers/registry.terraform.io/hashicorp/random/index.json'
+```
+
+### Sample Response
+
+```json
+{
+  "versions": {
+    "2.0.0": {},
+    "2.0.1": {}
+  }
+}
+```
+
+### Response Properties
+
+A successful result is a JSON object containing a single property `versions`,
+which must be a JSON object.
+
+Each of the property names of the `versions` object represents an available
+version number. The property values must be objects, but no properties are
+currently defined for those objects. Future versions of this protocol may
+define optional per-version properties for Terraform to use as installation
+hints, so implementations of the current version should leave those objects
+empty.
+
+Return `404 Not Found` to signal that the mirror does not have a provider
+with the given address.
+
+## List Available Installation Packages
+
+This operation returns download URLs of and associated metadata about the
+distribution packages for a particular version of a provider.
+
+Each distribution package is associated with a particular operating system
+and architecture. A network mirror may host only a subset of the available
+packages for a provider version, if the users of the mirror are known to all
+use only a subset of the target platforms that Terraform supports.
+
+Terraform CLI uses this operation after it has selected the newest available
+version matching the configured version constraints, in order to find a zip
+archive containing the plugin itself.
+
+| Method | Path                                       | Produces           |
+|--------|--------------------------------------------|--------------------|
+| `GET`  | `:hostname/:namespace/:type/:version.json` | `application/json` |
+
+### Parameters
+
+* `hostname` (required): the hostname portion of the address of the requested
+  provider.
+* `namespace` (required): the namespace portion of the address of the requested
+  provider.
+* `type` (required): the type portion of the address of the requested provider.
+* `version` (required): the version selected to download. This will exactly
```

Comment:
what do you think about "must" in place of "will"?

Reply:
This comment isn't clear, I meant in the last line, the `version` description: This _must_ exactly blah blah blah

Reply:
I think the oddness here is because this is written from the perspective of a potential server implementer rather than a potential _client_ implementer, so this is describing what the client (Terraform CLI) will do, rather than describing a requirement for the server implementer.

With that said, I do agree that it feels weird because protocol docs are normally written as instructions to future implementors rather than as descriptions of a particular implementation. I think I'm inclined to leave it as "will" for now, because other client implementations of this protocol feel unlikely to me, but maybe we could do another editorial pass later to ensure that the writing on this page is all written with a consistent audience in mind, because I expect there's also some holdovers here with content derived from the module registry docs, that were originally written as a description of how the _Terraform Registry_ behaves, because they started off as the registry's API docs.